packages/@aws-cdk/core/lib/asset-staging.ts

```diff
@@ -60,194 +104,244 @@ export class AssetStaging extends CoreConstruct {
 
   /**
-   * The path to the asset (stringinfied token).
+   * Absolute path to the asset data.
+   *
+   * If asset staging is disabled, this will just be the source path or
+   * a temporary directory used for bundling.
    *
-   * If asset staging is disabled, this will just be the original path.
    * If asset staging is enabled it will be the staged path.
    */
   public readonly stagedPath: string;
 
   /**
-   * The path of the asset as it was referenced by the user.
+   * The absolute path of the asset as it was referenced by the user.
    */
   public readonly sourcePath: string;
 
-  /**
-   * A cryptographic hash of the asset.
-   *
-   * @deprecated see `assetHash`.
-   */
-  public readonly sourceHash: string;
-
   /**
    * A cryptographic hash of the asset.
    */
   public readonly assetHash: string;
 
   private readonly fingerprintOptions: FingerprintOptions;
 
-  private readonly relativePath?: string;
+  private readonly hashType: AssetHashType;
+  private readonly outdir: string;
 
-  private bundleDir?: string;
+  /**
+   * A source source fingerprint given by the user
+   *
+   * Will not be used literally, always hashed later on.
+   */
+  private readonly customSourceFingerprint?: string;
 
   private readonly cacheKey: string;
 
   constructor(scope: Construct, id: string, props: AssetStagingProps) {
     super(scope, id);
 
-    this.sourcePath = props.sourcePath;
+    this.sourcePath = path.resolve(props.sourcePath);
     this.fingerprintOptions = props;
 
-    const outdir = Stage.of(this)?.outdir;
+    const outdir = Stage.of(this)?.assetOutdir;
     if (!outdir) {
       throw new Error('unable to determine cloud assembly output directory. Assets must be defined indirectly within a "Stage" or an "App" scope');
     }
+    this.outdir = outdir;
 
     // Determine the hash type based on the props as props.assetHashType is
     // optional from a caller perspective.
-    const hashType = determineHashType(props.assetHashType, props.assetHash);
+    this.customSourceFingerprint = props.assetHash;
+    this.hashType = determineHashType(props.assetHashType, this.customSourceFingerprint);
+
+    // Decide what we're going to do, without actually doing it yet
+    let stageThisAsset: () => StagedAsset;
+    let skip: boolean;
+    if (props.bundling) {
+      // Check if we actually have to bundle for this stack
+      const bundlingStacks: string[] = this.node.tryGetContext(cxapi.BUNDLING_STACKS) ?? ['*'];
+      skip = !bundlingStacks.find(pattern => minimatch(Stack.of(this).stackName, pattern));
+      const bundling = props.bundling;
+      stageThisAsset = () => this.stageByBundling(bundling, skip);
+    } else {
+      // "Staging disabled" only applies to copying. If we've already done the work for bundling,
+      // it's hardly any work to rename a directory afterwards.
+      skip = this.node.tryGetContext(cxapi.DISABLE_ASSET_STAGING_CONTEXT);
+      stageThisAsset = () => this.stageByCopying(skip);
```

Comment:
When staging is disabled and bundling is used, the staged path should be the (absolute) path to the bundling directory, is it still the case now?

Reply:
See https://github.com/aws/aws-cdk/issues/10367, users using this expect the `aws:asset:path` property of `Metadata` to be an absolute path to then bundling directory.

Reply:
Ah, good catch.

Staging disabled combined with bundling is not really a thing anymore.

It was not clear to me what the point of disabling staging was, because that wasn't explained in the docs or comments, so I assumed it was for performance reasons and concluded that if you've gone to the effort of bundling, one more or fewer `mv`s doesn't really make a difference.

If the point of "disabling staging" is to get an absolute path into `aws:asset:path`... well that seems a little silly. If `aws:asset:path` must always be an absolute path, why don't we then always stick an absolute path into it regardless? Why would it be relative if staging is enabled (thereby making it useless, apparently) and then have a flag to make it absolute?

Reply:
> It was not clear to me what the point of disabling staging was

Also not 100% clear to me, not using this feature. It was introduced to allow SAM to interact with CDK: https://docs.aws.amazon.com/cdk/latest/guide/sam.html

> If the point of "disabling staging" is to get an absolute path into `aws:asset:path`... well that seems a little silly.

Agree, may be the right solution as you say is to always output an absolute path. @eladb?

Reply:
But maybe we don't want to have absolute path in a deployed CF template.

Reply:
That's probably the reason. We always insert `aws:asset:path`, even though we're only interested in it if people also supply `--no-staging` (which is intended to be interpreted as `--integrate-with-sam-cli`).

Oh well. I'll clarify some more and fix the behavior. Thanks for the catch Jonathan!